### PR TITLE
Jr/editor font size setting

### PR DESCRIPTION
**Pull request by jeremy-rifkin**
Implements #3641, at least the core functionality.

### CODE REVIEW

File: static/settings.ts [hunk 1]
```diff
@@ -269,9 +279,10 @@ export class Settings {
             selector: string,
             name: keyof SiteSettings,
             populate: {label: string; desc: string}[],
-            defaultValue: string
+            defaultValue: any,
```
Comment (jeremy-rifkin) [inline static/settings.ts:282]: I'd like the type here to be `SiteSettings[name]` but I'm not sure if such a thing is possible in ts

Comment (RubenRBS) [reply static/settings.ts:282]: Using generics it should be possible, but then I don't know if the function would still be usabe in this context. See https://www.typescriptlang.org/play?#code/C4TwDgpgBAyhzAJYDsDmBnKBeKBvAsAFBQlQCGAXFMgK4C2ARhAE4A0RpUDV6wzKqdsVIBjKgwD2EgDYQyyIQF8A3ESKhIsRMAhwEAzDgLCSyMnQhUA1hBASAZrHhI06IZwAmEe2RrTgAGpk0jSWTvquANo2do56LhgAukSKaoT2NMgiSBLI5B4ecLLZEswAPABy5tAQAB46yB6YMQ7hCegAfAAUZhZUVRasUF4+foHBoVTxBpEDEIkAlHhpqYREZAVFECXMXQBEZHtDAIwL65sQxcCl+4dDe7V7C6qEG4WX29e7eyJHUHyhZ7nd5XG4-P60aTSZ5AA

Comment (jeremy-rifkin) [reply static/settings.ts:282]: Ah thank you! I think this can work

**Review by RubenRBS — APPROVED** [review]

File: static/colour.ts [hunk 2]
```diff
@@ -25,10 +25,12 @@
 import * as monaco from 'monaco-editor';
 import {Themes} from './themes';
 
+export type ColourScheme = 'rainbow' | 'rainbow2' | 'earth' | 'green-blue' | 'gray-shade' | 'rainbow-dark';
```
Comment (matsjla) [inline static/colour.ts:28]: Good catch

File: static/settings.ts [hunk 3]
```diff
@@ -265,13 +273,14 @@ export class Settings {
...
     private addSelectors() {
-        const addSelector = (
+        const addSelector = <Name extends keyof SiteSettings>(
             selector: string,
-            name: keyof SiteSettings,
+            name: Name,
             populate: {label: string; desc: string}[],
-            defaultValue: string
+            defaultValue: SiteSettings[Name],
+            component = Select
```
Comment (matsjla) [inline static/settings.ts:281]: Great stuff, love the use of the Name generic here

**Review by matsjla — APPROVED** [review]
LGTM - one thing I noticed is that the font size of existing editors doesn't change when the setting is changed, maybe this would be nice? I had to reset the layout for the change to apply

File: static/settings.ts (hunk 3 above)
Comment (jeremy-rifkin) [reply static/settings.ts:281]: Thanks to Ruben for showing me how to make this work!

**Comment by jeremy-rifkin** [comment]
> one thing I noticed is that the font size of existing editors doesn't change when the setting is changed, maybe this would be nice?

I was wondering about this too, I think I agree it'd be nice. I'll look into how to do it.

(One trick for now is just ctrl+click on all the font scale widgets to reset to the new default font scale)

**Comment by jeremy-rifkin** [comment]
The default font scale setting change is now emitted as an event so all panes that listen for it will automatically update. I've implemented it for pane.ts and compiler/editor.js. More panes will get this automatically as they're all converted to ts. Some panes like ir-view aren't calling the base class's register callbacks method at the moment.

I also added a button here to apply your default to all current panes in the case of a share link
image

**Comment by jeremy-rifkin** [comment]
I'll merge in the morning if all looks good

**Comment by jeremy-rifkin** [comment]
I'll merge in the morning if all looks good

**Review by RubenRBS — APPROVED** [review]

**Comment by partouf** [comment]
this is now live

**Pull request merged**